File: architecture-skills/design-patterns/examples/composite/composite.php
<?php
// Composite pattern in PHP.
//
// Files and directories implement one Node interface, so callers can
// total sizes or print trees without distinguishing leaf from branch.
// The classes carry a "Node" suffix because PHP already has a built-in
// Directory class.

interface Node
{
    public function size(): int;
    public function print(int $indent = 0): void;
}

class FileNode implements Node
{
    public function __construct(
        private string $name,
        private int $bytes,
    ) {}

    public function size(): int
    {
        return $this->bytes;
    }

    public function print(int $indent = 0): void
    {
        printf("%s%s (%d B)\n", str_repeat(' ', $indent), $this->name, $this->bytes);
    }
}

class DirectoryNode implements Node
{
    /** @var Node[] */
    private array $children = [];

    public function __construct(private string $name) {}

    public function add(Node $child): static
    {
        $this->children[] = $child;
        return $this;
    }

    public function size(): int
    {
        return array_sum(array_map(fn (Node $child) => $child->size(), $this->children));
    }

    public function print(int $indent = 0): void
    {
        printf("%s%s/ (%d B)\n", str_repeat(' ', $indent), $this->name, $this->size());
        foreach ($this->children as $child) {
            $child->print($indent + 2);
        }
    }
}

$project = (new DirectoryNode('project'))
    ->add((new DirectoryNode('src'))
        ->add(new FileNode('main.py', 1200))
        ->add(new FileNode('util.py', 450)))
    ->add((new DirectoryNode('docs'))->add(new FileNode('readme.md', 800)))
    ->add(new FileNode('.gitignore', 20));

$project->print();

File: architecture-skills/design-patterns/examples/composite/composite.cs
// Composite pattern in C#.
//
// Files and directories implement one INode interface, so callers can
// total sizes or print trees without distinguishing leaf from branch —
// the same shape as WPF's visual tree. Node classes carry a "Node"
// suffix to avoid clashing with System.IO.File/Directory.

using System;
using System.Collections.Generic;
using System.Linq;

interface INode
{
    long Size { get; }
    void Print(int indent = 0);
}

class FileNode : INode
{
    private readonly string _name;

    public FileNode(string name, long size)
    {
        _name = name;
        Size = size;
    }

    public long Size { get; }

    public void Print(int indent = 0) =>
        Console.WriteLine($"{new string(' ', indent)}{_name} ({Size} B)");
}

class DirectoryNode : INode
{
    private readonly string _name;
    private readonly List<INode> _children = new();

    public DirectoryNode(string name) => _name = name;

    public DirectoryNode Add(INode child)
    {
        _children.Add(child);
        return this;
    }

    public long Size => _children.Sum(child => child.Size);

    public void Print(int indent = 0)
    {
        Console.WriteLine($"{new string(' ', indent)}{_name}/ ({Size} B)");
        foreach (var child in _children)
        {
            child.Print(indent + 2);
        }
    }
}

class Program
{
    static void Main()
    {
        var project = new DirectoryNode("project")
            .Add(new DirectoryNode("src")
                .Add(new FileNode("main.py", 1200))
                .Add(new FileNode("util.py", 450)))
            .Add(new DirectoryNode("docs")
                .Add(new FileNode("readme.md", 800)))
            .Add(new FileNode(".gitignore", 20));
        project.Print();
    }
}
